service/src/test/java/com/commercetools/util/HttpRequestUtilParallelTest.java

```diff
@@ -24,52 +26,108 @@
+    }
+
+    /**
+     * Function to create request/response supplier for GET/POST tests and assert the received response. The result of
+     * this function is used as a supplier for {@link #makeAndAssertParallelRequests(int, int, Supplier)}.
+     * <p>
+     * The function implementation takes care about {@code 408 Timeout} response from
+     * <a href="http://httpbin.org">http://httpbin.org</a>: if such response is received
+     * (when the service is overloaded) - the implementation makes a retry up to 3 times. This is important to avoid
+     * false failures of the tests when <a href="http://httpbin.org">http://httpbin.org</a> is overloaded.
+     *
+     * @param responseSupplier       supplier which when executed makes single blocking request and returns a response.
+     * @param responseAssertConsumer procedure to assert the response obtained from {@code responseSupplier}
+     * @return new supplier which when finished successfully - returns <b>1</b>, otherwise - <b>0</b>.
+     */
+    private Supplier<Integer> createRequestsSupplier(final ExceptionalSupplier<HttpResponse> responseSupplier,
+                                                     final ExceptionalConsumer<HttpResponse> responseAssertConsumer) {
+        return () -> {
+            try {
+                int statusCode;
+                int repeatedTimes = 0;
+                HttpResponse httpResponse;
+                do {
+                    if (repeatedTimes > 0) {
+                        LOGGER.warn("{} \"Request Timeout\" received from [{}]. Retry {} of {} ...",
+                                SC_REQUEST_TIMEOUT, "http://httpbin.org/", repeatedTimes, RETRY_TIMES);
+                    }
+                    httpResponse = responseSupplier.get();
+                    statusCode = httpResponse.getStatusLine().getStatusCode();
+                    repeatedTimes++;
+                    // sometimes httpbin.org returns 408 (timeout) response. Note, this is not a InterruptedIOException
+                    // or ConnectException, thus DefaultHttpRequestRetryHandler re-try won't handle this case.
+                    // That's why in the tests we use this approach to re-try if the server responded successfully,
+                    // but with "408 timeout" response
+                } while (statusCode == SC_REQUEST_TIMEOUT && repeatedTimes <= RETRY_TIMES);
+
+                assertThat(repeatedTimes)
+                        .withFailMessage(format("Retries over-limit: the post request is executed %d times with timeout response %d",
+                                repeatedTimes, SC_REQUEST_TIMEOUT))
+                        .isLessThanOrEqualTo(RETRY_TIMES);
+
+                responseAssertConsumer.accept(httpResponse);
                 return 1;
-            } catch (IOException e) {
-                throw new RuntimeException("Request exception", e);
+            } catch (Exception e) {
+                LOGGER.error("Request exception: {}", e.toString());
```

Comment:
sometimes you use Strings.format and sometimes you use this string intepolation style. It's better if we are consistent with the style in the same code base. Otherwise, it feels like there is a reason behind it. Unless, there is a reason behind it here which I don't know 😄

Reply:
yes, there is explicit reason for this:
`slf4j` has it's own string substitution with postponed validation on demand (e.g. if logging actually happens), and this is exactly way recommended by `slf4j` developers:
https://www.slf4j.org/faq.html#logging_performance
https://www.slf4j.org/api/org/slf4j/helpers/MessageFormatter.html

in other cases i use `String.format`, which is actually just a formatting in-place, i.e. executed immediately.

Reply:
In the same time i noticed i make redundant `.toString()`, thanks: affc0a07